From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 17 Jan 2022 23:47:54 +0100
Subject: fix: crash when saving edited PDF files

This commit fixes a crash that persists any time a user attempts to
download an edited PDF. This was happening because the logic flow for
downloading of any edited PDF triggers a call to
chrome.fileSystem.chooseEntry, which we do not support and which
therefore causes unmapped page access crashes.

This patch can be removed should we choose to support chrome.fileSystem
or support it enough to fix the crash.

diff --git a/chrome/browser/resources/pdf/pdf_viewer.ts b/chrome/browser/resources/pdf/pdf_viewer.ts
index 2d28559d2eac827751d5dc82c127a4b967213499..8490f5118ef1c32fa5c3a8b010dca14a7928494b 100644
--- a/chrome/browser/resources/pdf/pdf_viewer.ts
+++ b/chrome/browser/resources/pdf/pdf_viewer.ts
@@ -1002,7 +1002,15 @@ export class PdfViewerElement extends PdfViewerBaseElement {
       dataArray = [result.dataToSave];
     }
 
+    const a = document.createElement('a');
+    a.download = this.attachments_[index].name;
     const blob = new Blob(dataArray);
+    // <if expr="not _google_chrome">
+    a.href = URL.createObjectURL(blob);
+    a.click();
+    URL.revokeObjectURL(a.href);
+    // </if>
+    // <if expr="_google_chrome">
     const fileName = this.attachments_[index].name;
     chrome.fileSystem.chooseEntry(
         {type: 'saveFile', suggestedName: fileName},
@@ -1024,6 +1032,7 @@ export class PdfViewerElement extends PdfViewerBaseElement {
             // </if>
           });
         });
+    // </if>
   }
 
   /**
@@ -1131,8 +1140,15 @@ export class PdfViewerElement extends PdfViewerBaseElement {
       fileName = fileName + '.pdf';
     }
 
-    // Create blob before callback to avoid race condition.
+    const a = document.createElement('a');
+    a.download = fileName;
     const blob = new Blob([result.dataToSave], {type: 'application/pdf'});
+    // <if expr="not _google_chrome">
+    a.href = URL.createObjectURL(blob);
+    a.click();
+    URL.revokeObjectURL(a.href);
+    // </if>
+    // <if expr="_google_chrome">
     chrome.fileSystem.chooseEntry(
         {
           type: 'saveFile',
@@ -1157,6 +1173,7 @@ export class PdfViewerElement extends PdfViewerBaseElement {
             // </if>
           });
         });
+    // </if>
 
     // <if expr="enable_ink">
     // Saving in Annotation mode is destructive: crbug.com/919364
